### ISSUE TITLE
ExecEngine Termination request

### ISSUE THREAD

**Issue by RieksJ** [AmpersandTarski/Ampersand#395]
Consider the following scriptlet:

     termScope :: Term * Scope [UNI,TOT]
     termCrit :: Term * Criterion [UNI]  REPRESENT Criterion TYPE ALPHANUMERIC

     varName :: Variable * Name [UNI,TOT]
     varScope :: Variable * Scope [UNI]
     varIsUsedBy :: Variable * Term

     RULE: varIsUsedBy;termScope |- varScope
     MEANING "When a term uses a variable, it must be in the same scope"

You can have the ExecEngine automatically create Variables when parsing the Criterion of a Term. 
Then, the ExecEngine can set the Scope of the Variable by inheriting it from the Term it was used in, as follows:

     ROLE ExecEngine MAINTAINS "Inherit Scope from Term"
     RULE "Inherit Scope from Term": varIsUsedBy;termScope |- VarScope
     VIOLATION (TXT "{EX} InsPair;varScope;Variable;", SRC I, TXT ";Term;", TGT I)

Consider the situation where the Variable is also (already) used by a Term in another Scope. Then, this rule would attempt to put the Variable in two Scopes, which would not succeed because `varScope` is UNI. Hence, the rule keeps firing until it ran in the limit of reruns (default: 10) that the ExecEngine does.

This case is an example of the more general case where the ExecEngine is instructed to solve violations that it san sometimes solve, but sometimes not. We can use rules to distinguish between these two situations. However, if we detect a case that the ExecEngine cannot solve, then there is no point in continuing execution, and it would be nice to have a mechanism that would 'kill' the ExecEngine runs from that point onwards.

So this is a request for an ExecEngine function that, when executed, terminates all further ExecEngine activity for the transaction in which it is called, using a syntax that will provide an appropriate error message.

**Comment by RieksJ** [comment]
Apparently, this issue is resolved by introducing a new ExecEngine function called `TerminateThisExecEngine($message)`, which takes one argument `$message` that specifies the text to be presented when it is invoked.

**Comment by Michiel-s** [comment]
> Apparently, this issue is resolved by introducing a new ExecEngine function called `TerminateThisExecEngine($message)`, which takes one argument `$message` that specifies the text to be presented when it is invoked.

That’s true. The message parameter is optional. When omitted the exec engine silently terminates. The log does contain information about where and when termination is invoked.

**Comment by Michiel-s** [comment]
@RieksJ, I think for the short term you are the only user of this feature. So let’s see how it fits purpose.
